addons/pos_mercury/static/src/js/PaymentScreen.js

```diff
@@ -110,6 +110,29 @@ odoo.define('pos_mercury.PaymentScreen', function (require) {
                 this.server_retries = 3;
             }
 
+            /**
+             * The card reader acts as a barcode scanner. This sets up
+             * the NumberBuffer to not immediately act on keyboard
+             * input.
+             *
+             * @override
+             */
+            get _getNumberBufferConfig() {
+                const res = super._getNumberBufferConfig;
+                res['useWithBarcode'] = true;
+                return res;
+            }
+
+            /**
+             * Finish any pending input before trying to validate.
+             *
+             * @override
+             */
+            async validateOrder(isForceValidate) {
+                NumberBuffer.capture();
+                return super.validateOrder(...arguments);
+            }
+
```

Comment:
Thanks for checking, I added it in response to these types of runbot errors ([more here](https://runbot.odoo.com/runbot/build/17181079)):

```
Error: Finalized Order cannot be modified
    at child.assert_editable (http://127.0.0.1:8069/web/content/2010-eefa310/point_of_sale.assets.js:525:127)
    at child.set_amount (http://127.0.0.1:8069/web/content/2010-eefa310/point_of_sale.assets.js:508:490)
    at PaymentScreen._updateSelectedPaymentline (http://127.0.0.1:8069/web/content/2010-eefa310/point_of_sale.assets.js:991:130)
    at PaymentScreen.__trigger (http://127.0.0.1:8069/web/content/2008-2633e56/web.assets_common.js:1268:9)
    at PaymentScreen.owl.Component.__trigger (http://127.0.0.1:8069/web/content/2008-2633e56/web.assets_common.js:1420:17)
    at PaymentScreen.trigger (http://127.0.0.1:8069/web/content/2008-2633e56/web.assets_common.js:1257:33)
    at NumberBuffer._handleInput (http://127.0.0.1:8069/web/content/2010-eefa310/point_of_sale.assets.js:1445:16)
    at http://127.0.0.1:8069/web/content/2010-eefa310/point_of_sale.assets.js:1442:42
```

It's reproduceable locally as well if you're fast enough:

0/ go to the payment screen
1/ click cash (auto-fills the right amount)
2/ type some numbers
3/ immediately click validate (before the `NumberBuffer` calls its handler)

If the PaymentScreen is gone when the [NumberBuffer handler](https://github.com/odoo/odoo/blob/939ffdfadd82789d8ad17245f48814270755b042/addons/point_of_sale/static/src/js/Misc/NumberBuffer.js#L181) runs we're fine because [OWL ignores events on components without an el](https://github.com/odoo/odoo/blob/939ffdfadd82789d8ad17245f48814270755b042/addons/web/static/lib/owl/owl.js#L4424). But before advancing to the receipt screen we first wait on the order to sync. [_finalizeValidation()](https://github.com/odoo/odoo/blob/939ffdfadd82789d8ad17245f48814270755b042/addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentScreen.js#L174) will set `finalized` and then [await the sync](https://github.com/odoo/odoo/blob/939ffdfadd82789d8ad17245f48814270755b042/addons/point_of_sale/static/src/js/Screens/PaymentScreen/PaymentScreen.js#L184). If the NumberBuffer handler runs during this time you end up with the error.

> perhaps we should call the NumberBuffer.capture in the point_of_sale addon.

Theoretically there's a race condition but I don't think it's causing issues with the default `setTimeout(..., 0)` in `NumberBuffer`. I'm not sure  we want to risk fixing something that isn't broken, especially since this is an older version and nobody seemed to have run into the issue.

Reply:
Thanks for the detailed explanation. I wanted to see the error in action myself but while fetching your changes, the explanation clicked. I also agree with your reasoning on introducing the "capture" here. I'll investigate in master if "capture" call is needed in point_of_sale addon.